cpp/src/fil/infer.cu

```diff
@@ -72,10 +86,122 @@ __device__ __forceinline__ void infer_one_tree(tree_type tree, float* sdata,
     bool cond = isnan(val) ? !n.def_left() : val >= n.thresh();
     curr = n.left(curr) + cond;
   }
-  out[0] = tree[curr].output();
+  vec<1, output_type> out;
+  out[0] = tree[curr].base_node::output<output_type>();
+  return out;
 }
 
-template <int NITEMS, typename storage_type>
+template <int NITEMS,
+          leaf_value_t leaf_payload_type>  // = FLOAT_SCALAR
+struct tree_aggregator_t {
+  /** To compute accurately, would need to know the latest __CUDA_ARCH__
```

Comment:
We don't need to pass the `ptx_arch` explicitly.

Reply:
couldn't think of anything better than launch an init kernel with the sole purpose to compute the footprint (otherwise, we're back to static shared memory)
from algorithm design perspective, I don't see why CUB shared memory footprint should be/stay low, so wondering if you're OK with this approach?

Reply:
so far, it's +20 lines of code, and it would launch 4 additional test kernels on every inference.
there are many directions that we can take from here, but the main question is - if the patch I already uploaded is more safe than the code in 0.13/0.14 branches (that assumes 0 footprint for CUB), why not address this properly in another PR?

Reply:
decided to properly fix in another PR. as-is, it's already safer than the current code, because it estimates a non-zero footprint for CUB, and the CUB memory is now co-located with input data row memory, making it more likely that max (up to 4) items/block will exceed CUB footprint